### PR TITLE
Add support for aggregates on collections (#807)

### PR DESCRIPTION
**Pull request by ashwinphatak**
## What, How & Why?

Reference to the issue(s) addressed by this PR: # ???

## ☑️ ToDos

* [x] 📝 Changelog entry
* [x] 🚦 Tests
* [x] 📝 Public documentation PR created or is not necessary
* [x] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [x] typescript definitions file is updated
* [x] jsdoc files updated
* [x] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

**Comment by ashwinphatak** [comment]
@kneth these changes are from your branch, rebased on master with merge conflicts removed. Will continue from this point.

**Comment by bmunkholm** [comment]
@ashwinphatak We prefer not to represent "state" in the title with e.g. "WIP". We use the Pipeline in Zenhub instead.

**Comment by ashwinphatak** [comment]
@bmunkholm Done. I wasn't using Zendesk earlier as it wouldn't show me realm-js in my list of repos on signing up. I'd presumed (incorrectly) that it required access to the Realm GitHub org, but it turns out one has to search for public repositories explicitly.

File: src/js_list.hpp [hunk 1]
```diff
@@ -116,6 +126,75 @@ void ListClass<T>::get_length(ContextType, ObjectType object, ReturnValue &retur
...
+void ListClass<T>::max(ContextType ctx, FunctionType, ObjectType this_object, size_t argc, const ValueType arguments[], ReturnValue &return_value) {
+    validate_argument_count_at_least(argc, 1);
+    std::string property_name = Value::validated_to_string(ctx, arguments[0]);
+
+    auto list = get_internal<T, ListClass<T>>(this_object);
+
+    const ObjectSchema& object_schema = list->get_object_schema();
+    const Property* property = object_schema.property_for_name(property_name);
+    if (!property) {
+        throw std::invalid_argument("No such property");
```
Comment (kneth) [inline src/js_list.hpp]: Can we have the same error message as in line 138?

File: src/js_list.hpp [hunk 2]
```diff
@@ -116,6 +126,75 @@ void ListClass<T>::get_length(ContextType, ObjectType object, ReturnValue &retur
...
+void ListClass<T>::sum(ContextType ctx, FunctionType, ObjectType this_object, size_t argc, const ValueType arguments[], ReturnValue &return_value) {
+    validate_argument_count_at_least(argc, 1);
+    std::string property_name = Value::validated_to_string(ctx, arguments[0]);
+
+    auto list = get_internal<T, ListClass<T>>(this_object);
+
+    const ObjectSchema& object_schema = list->get_object_schema();
+    const Property* property = object_schema.property_for_name(property_name);
+    if (!property) {
+        throw std::invalid_argument("No such property");
```
Comment (kneth) [inline src/js_list.hpp]: Can we have the same error message as in line 138?

File: src/js_list.hpp [hunk 3]
```diff
@@ -116,6 +126,75 @@ void ListClass<T>::get_length(ContextType, ObjectType object, ReturnValue &retur
...
+void ListClass<T>::avg(ContextType ctx, FunctionType, ObjectType this_object, size_t argc, const ValueType arguments[], ReturnValue &return_value) {
+    validate_argument_count_at_least(argc, 1);
+    std::string property_name = Value::validated_to_string(ctx, arguments[0]);
+
+    auto list = get_internal<T, ListClass<T>>(this_object);
+
+    const ObjectSchema& object_schema = list->get_object_schema();
+    const Property* property = object_schema.property_for_name(property_name);
+    if (!property) {
+        throw std::invalid_argument("No such property");
```
Comment (kneth) [inline src/js_list.hpp]: Can we have the same error message as in line 138?

**Comment by ashwinphatak** [comment]
@kneth good point, I'll try to extract the logic into a common helper function.

**Comment by ashwinphatak** [comment]
@kneth extracted a common helper. Could you also please restart CI?

File: lib/index.d.ts [hunk 4]
```diff
@@ -141,6 +141,34 @@ declare namespace Realm {
         isValid(): boolean;
 
         /**
+         * Computes the minimum value.
+         * @param  {string} property_name
+         * @returns number
+         */
+        min(property_name: string): number;
```
Comment (blagoev) [inline lib/index.d.ts]: can we name this `property` to be on par with the other methods

Comment (blagoev) [inline lib/index.d.ts]: can we name this `property` to be on par with the other methods

File: src/js_list.hpp [hunk 5]
```diff
@@ -47,6 +47,7 @@ class List : public realm::List {
 
 template<typename T>
 struct ListClass : ClassDefinition<T, realm::js::List<T>, CollectionClass<T>> {
+    using Type = T;
```
Comment (blagoev) [inline src/js_list.hpp]: this seems unused

File: src/js_results.hpp [hunk 6]
```diff
@@ -49,6 +50,7 @@ class Results : public realm::Results {
 
 template<typename T>
 struct ResultsClass : ClassDefinition<T, realm::js::Results<T>, CollectionClass<T>> {
+    using Type = T;
```
Comment (blagoev) [inline src/js_results.hpp]: seems not used

File: src/js_list.hpp (hunk 5 above)
Comment (ashwinphatak) [reply src/js_list.hpp]: It's used inside `compute_aggregate_on_collection`: `get_internal<typename T::Type, T>(this_object)`

File: src/js_results.hpp (hunk 6 above)
Comment (ashwinphatak) [reply src/js_results.hpp]: It's used inside `compute_aggregate_on_collection`: `get_internal<typename T::Type, T>(this_object)`

**Pull request merged**